### PR TITLE
Preserve credentials but detroy session on logout

### LINKED ISSUE #1098: [Quest] Enhance Authentication to support offline subsequent login
**Context**
- All FHIR Core apps require to have offline capabilities. This means all subsequent logins after the initial login should be offline capable. 
- Right now we save more of the configuration and authentication data in the secure shared preferences. 
- The current logout clears out the secure shared preferences. This means after the logout a user cannot login offline. 

**Feature implementation** 
- [x] Update the logout functionality to not clear the secure share preference. 
- [x] Enable login from the shared preferences after logout.  
- [x] Enable accessing the settings/configurations page via the login and pin logins pages.  
    - [x] In case a user change the composition Id or app Id clear all the shared preferences and retrigger the configuration.
    - [x] After clicking save on the setting/configuration page 
        - [x]  If offline,  show a warning dialog warning them that the share preferences will be cleared and they would not be able to login offline again.  If they click yes proceed to clear the shared preferences and log out. If they click no do not save the changes. 
        - [x]  If online,  show a dialog asking them whether they are sure of the operation, If they click yes proceed to clear the shared preferences and log out. If they click no do not save the changes. 

**Acceptance criteria**
- [x] Ability to login offline for all subsequent logins
- [ ] Clearing all data after settings update on the settings page.

### PR DESCRIPTION
**Pull request by maimoonak**
Fixes part of #1098 

**Type**
Choose one: Feature

**Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [x] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [x] I have built and run the fhircore app to verify my change fixes the issue and/or does not break the app

### CODE REVIEW

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

**Review by dubdabasoduba — COMMENTED** [review]
Getting this error which also breaks sync. 
```
2022-03-08 16:15:53.596 3343-3343/org.smartregister.fhircore.anc W/AndroidKeysetManager: keyset not found, will generate a new one
    java.io.FileNotFoundException: can't read keyset; the pref value __androidx_security_crypto_encrypted_prefs_key_keyset__ does not exist
        at com.google.crypto.tink.integration.android.SharedPrefKeysetReader.readPref(SharedPrefKeysetReader.java:71)
        at com.google.crypto.tink.integration.android.SharedPrefKeysetReader.readEncrypted(SharedPrefKeysetReader.java:89)
        at com.google.crypto.tink.KeysetHandle.read(KeysetHandle.java:105)
        at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.read(AndroidKeysetManager.java:311)
        at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.readOrGenerateNewKeyset(AndroidKeysetManager.java:287)
        at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.build(AndroidKeysetManager.java:238)
        at androidx.security.crypto.EncryptedSharedPreferences.create(EncryptedSharedPreferences.java:155)
        at androidx.security.crypto.EncryptedSharedPreferences.create(EncryptedSharedPreferences.java:120)
        at org.smartregister.fhircore.engine.util.SecureSharedPreference.<init>(SecureSharedPreference.kt:34)
        at org.smartregister.fhircore.anc.DaggerAncApplication_HiltComponents_SingletonC.secureSharedPreference(DaggerAncApplication_HiltComponents_SingletonC.java:233)
        at org.smartregister.fhircore.anc.DaggerAncApplication_HiltComponents_SingletonC.access$5300(DaggerAncApplication_HiltComponents_SingletonC.java:150)
        at org.smartregister.fhircore.anc.DaggerAncApplication_HiltComponents_SingletonC$SwitchingProvider.get(DaggerAncApplication_HiltComponents_SingletonC.java:1301)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
        at org.smartregister.fhircore.anc.DaggerAncApplication_HiltComponents_SingletonC.tokenManagerService(DaggerAncApplication_HiltComponents_SingletonC.java:237)
        at org.smartregister.fhircore.anc.DaggerAncApplication_HiltComponents_SingletonC.access$5100(DaggerAncApplication_HiltComponents_SingletonC.java:150)
        at org.smartregister.fhircore.anc.DaggerAncApplication_HiltComponents_SingletonC$SwitchingProvider.get(DaggerAncApplication_HiltComponents_SingletonC.java:1295)
        at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
````

**How to replicate.** 
- Using the `1098_logout_fix` branch build the app. 
- Login using `appId`: `ecbis` 
- Set up pin. 
- You will be redirected to the register, the sync dialog will appear but if you look at the logcat logs sync is not happening. The above error is throw instead. 

**Test device specs**
- Android 11 emulator

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/remote/shared/interceptor/OAuthInterceptor.kt [hunk 1]
```diff
@@ -47,7 +47,9 @@ constructor(
   fun Request.hasPaths(paths: List<String>) = this.url.pathSegments.containsAll(paths)
 
   fun Request.hasOpenResources() =
-    this.url.pathSegments.any {
-      it.contentEquals(ResourceType.Composition.name) || it.contentEquals(ResourceType.Binary.name)
-    }
+    this.method.contentEquals("GET") &&
```
[the remark this answers is not in the document]
Comment (maimoonak) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/data/remote/shared/interceptor/OAuthInterceptor.kt]: done

**Comment by maimoonak** [comment]
@dubdabasoduba the sync issue was due to main branch which yesterday became unstable due to some merges. It should work now

**Comment by dubdabasoduba** [comment]
@maimoonak still not able to sync on this PR.

**Comment by maimoonak** [comment]
@dubdabasoduba The issue was that PIN setup/login was not running sync. I have resolved these and tested as well. @RaaziaTarique @AbdulWahabMemon plz test all 6 apps once again and update on this PR.

**Review by dubdabasoduba — COMMENTED** [review]
The app crashes on logout while offline. 
**How to replicate**
- Login to the app using the `ecbis` appId. 
- Let the app complete sync. 
- Put your device on Airplane mode and then hit logout. 
- The app crashes.  

Acceptance criteria
- [x] Login to the app using the `ecbis` appId. 
- [x] Let the app complete sync. 
- [ ]  Put your device on Airplane mode and then hit logout. 
- [ ] The app should log out gracefully without crashes.

**Comment by dubdabasoduba** [comment]
> @dubdabasoduba The issue was that PIN setup/login was not running sync. I have resolved these and tested as well. @RaaziaTarique @AbdulWahabMemon plz test all 6 apps once again and update on this PR.

Sync was fixed @maimoonak

**Comment by RaaziaTarique** [comment]
> The app crashes on logout while offline. **How to replicate**
> 
> * Login to the app using the `ecbis` appId.
> * Let the app complete sync.
> * Put your device on Airplane mode and then hit logout.
> * The app crashes.
> 
> Acceptance criteria
> 
> * [ ]  Login to the app using the `ecbis` appId.
> * [ ]  Let the app complete sync.
> * [ ]   Put your device on Airplane mode and then hit logout.
> * [ ]  The app should log out gracefully without crashes.

I have looked into this issue and I have found a fix. Should I commit the changes into this PR?

**Comment by dubdabasoduba** [comment]
> > The app crashes on logout while offline. **How to replicate**
> > 
> > * Login to the app using the `ecbis` appId.
> > * Let the app complete sync.
> > * Put your device on Airplane mode and then hit logout.
> > * The app crashes.
> > 
> > Acceptance criteria
> > 
> > * [ ]   Login to the app using the `ecbis` appId.
> > * [ ]   Let the app complete sync.
> > * [ ]    Put your device on Airplane mode and then hit logout.
> > * [ ]   The app should log out gracefully without crashes.
> 
> I have looked into this issue and I have found a fix. Should I commit the changes into this PR?

Yes please.

**Comment by dubdabasoduba** [comment]
@RaaziaTarique I think we have now introduced a regression. 

- [ ] On logout if the PIN is set the user should be redirected to the PIN Login page. 
- [ ] On logout if the PIN is not set the user should be redirected to the username.password page. 
- [ ] Both set pin and pin login pages should have a menu name "User login" that redirects to the username/password login
- [ ] If a user redirects to the username/password login page and had PIN set up earlier they should be redirected to the main register after login. Currently, the app forces the user to set the PIN again. 

Please test all these scenarios correctly during feature development.

**Comment by RaaziaTarique** [comment]
> @RaaziaTarique I think we have now introduced a regression.
> 
> * [x]  On logout if the PIN is set the user should be redirected to the PIN Login page.
> * [x]  On logout if the PIN is not set the user should be redirected to the username.password page.
> * [ ]  Both set pin and pin login pages should have a menu name "User login" that redirects to the username/password login
> * [x]  If a user redirects to the username/password login page and had PIN set up earlier they should be redirected to the main register after login. Currently, the app forces the user to set the PIN again.
> 
> Please test all these scenarios correctly during feature development.
> 
> #1070 #914 #912 #911
> 
> All the issues tagged here have the fixes mentioned above. Please find out where we reverted the desired functionality.

Hi @dubdabasoduba, related to point 3. Just wanted to make sure one thing, if user goes back to user/password page from Pin setup then the user should be directed to home page directly. CC: @AbdulWahabMemon

File: android/engine/src/main/java/org/smartregister/fhircore/engine/ui/pin/PinLoginActivity.kt [hunk 2]
```diff
@@ -61,10 +64,12 @@ class PinLoginActivity : BaseMultiLanguageActivity() {
   }
 
   private fun moveToHome() {
+    syncBroadcaster.runSync()
     loginService.navigateToHome()
   }
 
   private fun moveToLoginViaUsername() {
+    Timber.i("login 2")
```
Comment (dubdabasoduba) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/ui/pin/PinLoginActivity.kt:72]: Do we need this log? If yes, please update it to a more informative one

**Review by dubdabasoduba — CHANGES_REQUESTED** [review]

**Comment by dubdabasoduba** [comment]
@RaaziaTarique I tested this. It works well.  Just one comment and we merge it. Good job.

File: android/engine/src/main/java/org/smartregister/fhircore/engine/ui/pin/PinLoginActivity.kt (hunk 2 above)
Comment (RaaziaTarique) [reply android/engine/src/main/java/org/smartregister/fhircore/engine/ui/pin/PinLoginActivity.kt:72]: @dubdabasoduba No, we don't need this log. I added it for testing. Removing it.

**Review by dubdabasoduba — APPROVED** [review]

**Pull request merged**